basalt-server-lib/src/server/teams.rs

```diff
@@ -0,0 +1,193 @@
+use bedrock::Config;
+use chrono::Utc;
+use dashmap::DashMap;
+use serde::{Deserialize, Serialize};
+
+use crate::repositories::users::Username;
+
+#[derive(Debug, PartialEq, Eq, Default, Copy, Clone, Deserialize, Serialize, utoipa::ToSchema)]
+#[serde(rename_all = "camelCase")]
+pub struct TeamInfo {
+    /// When the team last contacted the server
+    pub last_seen: Option<chrono::DateTime<Utc>>,
+    /// Whether or not the team has checked into the competition by logging in
+    pub checked_in: bool,
+    /// Just a flag stating whether or not the team has deliberately disconnected
+    pub disconnected: bool,
+}
+
+#[derive(Debug, PartialEq, Eq, Clone, Deserialize, Serialize, utoipa::ToSchema)]
+#[serde(rename_all = "camelCase")]
+pub struct TeamFull {
+    /// Username of team/player
+    pub team: Username,
+    /// Contains full information about team
+    #[serde(flatten)]
+    pub info: TeamInfo,
+}
+
+impl TeamInfo {
+    fn check(&mut self) {
+        self.checked_in = true;
+        self.last_seen = Some(Utc::now());
+        self.disconnected = false;
+    }
+    fn disconnect(&mut self) {
+        self.disconnected = true;
+    }
+}
+
+#[derive(Clone, Debug, PartialEq, Deserialize, serde::Serialize, utoipa::ToSchema)]
+pub struct TeamWithScore {
+    pub score: f64,
+    #[serde(flatten)]
+    pub team_info: TeamFull,
+}
+
+pub struct TeamManagement {
+    teams: DashMap<Username, TeamInfo>,
+}
+
+impl TeamManagement {
+    pub fn from_config(cfg: &Config) -> Self {
+        let teams: DashMap<Username, TeamInfo> = DashMap::new();
+        for t in &cfg.accounts.competitors {
+            teams.insert(t.name.clone().into(), TeamInfo::default());
+        }
+        TeamManagement { teams }
+    }
+
+    pub fn check_in(&self, name: &Username) {
+        self.teams.entry(name.clone()).and_modify(|t| t.check());
+    }
+
+    pub fn disconnect(&self, name: &Username) {
+        self.teams
+            .entry(name.clone())
+            .and_modify(|t| t.disconnect());
+    }
```

Comment:
Using the entry api here leads to unnecessary clones.  It would be better to just use `.get_mut` and if-let:

```suggestion
    pub fn check_in(&self, name: &Username) {
        if let Some(mut t) = self.teams.get_mut(name) {
            t.check()
        }
    }

    pub fn disconnect(&self, name: &Username) {
        if let Some(mut t) = self.teams.get_mut(name) {
            t.disconnect();
        }
    }
```